README.md
# n32s

The n32s code is a precompiler from Notation3 (RDF Surfaces) to a N3S format that is used by Prolog reasoners such as [eye](https://github.com/eyereasoner/eye), [retina](https://github.com/eyereasoner/retina) and [Latar](https://github.com/MellonScholarlyCommunication/Latar).

## Dependency

Need node version >= 18.

Need N3.js version 2.0.0

```
git clone git@github.com:rdfjs/N3.js.git
git fetch --all
git checkout versions/2.0.0
cd N3.js
yarn 
yarn link
```

## Installation

```
yarn install
yarn link n3
yarn build
```

## Usage

```
node js/index.js <n3-file>
```

## Example

```
$ cat demo.n3
@prefix : <urn:example.org:> .
@prefix log: <http://www.w3.org/2000/10/swap/log#> .

:Alice a :Person .

(_:X) log:onNegativeSurface {
    _:X a :Person .
    () log:onNegativeSurface {
        _:X a :Human .
    } .
} .

$ node js/index.js demo.n3
'<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('<urn:example.org:Alice>','<urn:example.org:Person>').
'<http://www.w3.org/2000/10/swap/log#onNegativeSurface>'(['_:X'],('<http://www.w3.org/2000/10/swap/log#onNegativeSurface>'([],('<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('_:X','<urn:example.org:Human>'))),'<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('_:X','<urn:example.org:Person>'))).
```

Changes
2023-06-22
   - Fixing graph list issues using N3.js v.2.0.0
  
2023-06-19
   - Hardening the ordering bug fix
   - Fixing string literal processing
  
2023-06-18 
   - Fixing empty list bug (obs. josd)
   - Fixing ordering bug (obs. josd)
   - Starting with a N3S lexer and parser

2023-06-17 0.0.4
   - Fixing dynamics
  
2023-06-17 0.0.3
   - Adding dynamics
   - Fixing literals
   - Removing support for ?x variables
   - Attempt to parse broken lists

2023-06-16 0.0.2
   - Supporting variables
   - Supporting literals in subject position
   
2023-06-16 0.0.1
   - Initial release

test.sh
#!/bin/bash

for f in data/*.n3; do
    echo "[$f]"
    node js/index.js $f | tee ${f}-result
    echo "---"
done


test/N3Parser.spec.ts
import { N3Parser } from '../src/N3Parser';

import {expect} from 'expect';

describe('N3Parser', () => {
    it('should parse the empty string',
        shouldParse('','')
    );

    it('should parse <s> <p> <o>',
        shouldParse(
            '<s> <p> <o>.',
            '\'<p>\'(\'<s>\',\'<o>\').')
    );

    it('should parse <s> <p> true',
        shouldParse(
            '<s> <p> true.',
            '\'<p>\'(\'<s>\',true).')
    );

    it('should parse <s> <p> false',
        shouldParse(
            '<s> <p> false.',
            '\'<p>\'(\'<s>\',false).')
    );

    it('should parse <s> <p> "blue ball"',
        shouldParse(
            '<s> <p> "blue ball".',
            '\'<p>\'(\'<s>\',"blue ball").')
    );

    it('should parse <s> <p> 123',
        shouldParse(
            '<s> <p> 123.',
            '\'<p>\'(\'<s>\',123).')
    );

    it('should parse <s> <p> "3.14"^^<http://www.w3.org/2001/XMLSchema#float>',
        shouldParse(
            '<s> <p> "3.14"^^<http://www.w3.org/2001/XMLSchema#float>.',
            '\'<p>\'(\'<s>\',literal(\'3.14\',\'http://www.w3.org/2001/XMLSchema#float\')).')
    );

    it('should parse <s> <p> _:x',
        shouldParse(
            '<s> <p> _:x.',
            '\'<p>\'(\'<s>\',\'_:x\').')
    );

    it('should parse <s> <p> ()',
        shouldParse(
            '<s> <p> ().',
            '\'<p>\'(\'<s>\',[]).')
    );

    it('should parse <s> <p> (())',
        shouldParse(
            '<s> <p> (()).',
            '\'<p>\'(\'<s>\',[[]]).')
    );

    it('should parse <s> <p> (() ())',
        shouldParse(
            '<s> <p> (() ()).',
            '\'<p>\'(\'<s>\',[[],[]]).')
    );

    it('should parse <s> <p> (() "#" <x> (<y>))',
        shouldParse(
            '<s> <p> (() "#" <x> (<y>)).',
            '\'<p>\'(\'<s>\',[[],"#",\'<x>\',[\'<y>\']]).')
    );

    it('should parse <s> <p> { <a> <b> <c> }',
        shouldParse(
            '<s> <p> { <a> <b> <c> }.',
            '\'<p>\'(\'<s>\',(\'<b>\'(\'<a>\',\'<c>\'))).')
    );
});

function shouldParse(input:string,expected:string) {
    return async function() {
        const result = await makeN3S(input);
        expect(result).toEqual(expected);
    };
}
async function makeN3S(input: string) : Promise<string> {
    const parser = new N3Parser({ dynamics: false });
    const graph = await parser.parse(input);
    const n3s = parser.asN3S(graph);
    return n3s;
}

test/N3SLexer.spec.ts
import { N3SLexer }  from '../src/N3SLexer';
import {expect} from 'expect';

describe('N3SLexer', () => {
  describe('The N3SLexer export', () => {
    it('should be an N3SLexer constructor', () => {
      const lexer = new N3SLexer();
      expect(lexer).not.toBeNull();
    });
  });

  describe('A N3SLexer instance', () => {
    function createLexer() { return new N3SLexer({comments: true}); }
    it('should tokenize the empty string',
      shouldTokenize(createLexer(),
                    '',
                    [{ type: 'eof', line: 0 }]));

    it('should tokenize an IRI',
      shouldTokenize(createLexer(),
                    '\'<urn:foo>\'',
                    [ { type: 'IRI', value: 'urn:foo', line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a blank node',
      shouldTokenize(createLexer(),
                    '\'_:X\'',
                    [ { type: 'blank', value: 'X', prefix: '_' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize the boolean true',
      shouldTokenize(createLexer(),
                    'true',
                    [ { type: 'literal', value: 'true', prefix: 'http://www.w3.org/2001/XMLSchema#boolean' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize the boolean false',
      shouldTokenize(createLexer(),
                    'false',
                    [ { type: 'literal', value: 'false', prefix: 'http://www.w3.org/2001/XMLSchema#boolean' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a string value',
      shouldTokenize(createLexer(),
                    '\'a string\'',
                    [ { type: 'literal', value: 'a string', prefix: 'http://www.w3.org/2001/XMLSchema#string' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize an integer value',
      shouldTokenize(createLexer(),
                    '42',
                    [ { type: 'literal', value: '42', prefix: 'http://www.w3.org/2001/XMLSchema#integer' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a negative integer value',
      shouldTokenize(createLexer(),
                    '-42',
                    [ { type: 'literal', value: '-42', prefix: 'http://www.w3.org/2001/XMLSchema#integer' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a decimal value',
      shouldTokenize(createLexer(),
                    '42.1',
                    [ { type: 'literal', value: '42.1', prefix: 'http://www.w3.org/2001/XMLSchema#decimal' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a literal value',
      shouldTokenize(createLexer(),
                    'literal(42,\'http://www.w3.org/2001/XMLSchema#integer\')',
                    [ { type: 'literal', value: '42', prefix: 'http://www.w3.org/2001/XMLSchema#integer' , line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a list value',
      shouldTokenize(createLexer(),
                    '[\'_:x\',\'_:y\']',
                    [ { type: '[', line: 0 } ,
                      { type: 'blank', value: 'x', prefix: '_' , line: 0 } ,
                      { type: ',', line: 0 } ,
                      { type: 'blank', value: 'y', prefix: '_' , line: 0 } ,
                      { type: ']', line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a conjunction value',
      shouldTokenize(createLexer(),
                    '(\'_:x\',\'_:y\')',
                    [ { type: '(', line: 0 } ,
                      { type: 'blank', value: 'x', prefix: '_' , line: 0 } ,
                      { type: ',', line: 0 } ,
                      { type: 'blank', value: 'y', prefix: '_' , line: 0 } ,
                      { type: ')', line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a directive',
      shouldTokenize(createLexer(),
                    ':- dynamic(\'<urn:example.org:value>\'/2).',
                    [ { type: 'directive', line: 0 , value: 'dynamic(\'<urn:example.org:value>\'/2).'} ,
                      { type: 'eof', line: 0 }
                    ]));

    it('should tokenize a comment',
      shouldTokenize(createLexer(),
                    "% a comment\n",
                    [ { type: 'comment', line: 0 , value: ' a comment'} ,
                      { type: 'eof', line: 1 }
                    ]));

    it('should tokenize the end of term',
      shouldTokenize(createLexer(),
                    '.',
                    [ { type: '.', line: 0 } ,
                      { type: 'eof', line: 0 }
                    ]));
  });
});

function shouldTokenize(lexer: N3SLexer, input: string, expected: any) {
  const ignoredAttributes : any = { start: true, end: true };

  return function (done: any) {
    const result : any[] = [];
    lexer.tokenize(input, tokenCallback);

    function tokenCallback(error: Error | null, token: any) {
      expect(error).toBeNull();
      expect(token).not.toBeNull();
      const expectedItem = expected[result.length];
      if (expectedItem) {
        for (const attribute in token) {
          if (typeof expectedItem[attribute] === 'undefined' &&
              (token[attribute] === '' || (ignoredAttributes[attribute]))) {
            delete token[attribute];
          }
        }
      }
      result.push(token);
      if (token.type === 'eof') {
        expect(result).toEqual(expected);
        done(null, result);
      }
    }
  };
}

src/index.ts
#!/usr/bin/env node

import { N3Parser } from './N3Parser';
export { N3Parser } from './N3Parser';
import { program } from 'commander';
import * as log4js from 'log4js';

program.version('0.0.4')
       .argument('<file>')
       .option('-d,--info','output debugging messages')
       .option('-dd,--debug','output more debugging messages')
       .option('-ddd,--trace','output much more debugging messages');

program.parse(process.argv);

const opts   = program.opts();
const logger = log4js.getLogger();

if (opts.info) {
  logger.level = "info";
}

if (opts.debug) {
  logger.level = "debug";
}

if (opts.trace) {
  logger.level = "trace";
}

const input = program.args[0];

logger.info(`input: ${input}`);

const knownPredicates = [
    'http://www.w3.org/1999/02/22-rdf-syntax-ns#type',
    'http://www.w3.org/2000/10/swap/.*',
];

main(input);

async function main(path: string) : Promise<void> {
    const parser = new N3Parser({});
    parser.parse(path)
          .then(graph => {
            const n3s = parser.asN3S(graph, knownPredicates);
            console.log(n3s);
          })
          .catch(e => {
            console.error(e);
            process.exit(2);
          });
}

src/parser.ts
import * as N3 from 'n3';
import * as fs from 'fs';
import { getLogger } from "log4js";

const XSD = 'http://www.w3.org/2001/XMLSchema#';
const RDF = 'http://www.w3.org/1999/02/22-rdf-syntax-ns#';

const logger = getLogger();

export type IPSO = {
    type: 'PSO',
    subject: ITerm ,
    predicate: ITerm ,
    object: ITerm ,
};

export type ITerm = INamedNode | IBlankNode | ILiteral | IVariable | IGraph | IList;

export type INamedNode = {
    type: 'NamedNode';
    value: string;
    datatype: null;
};

export type IBlankNode = {
    type: 'BlankNode';
    value: string;
    datatype: null;
};

export type ILiteral = {
    type: 'Literal';
    value: string;
    datatype: string;
};

export type IVariable = {
    type: 'Variable';
    value: string;
    datatype: null;
};

export type IList = {
    type: 'List';
    value: ITerm[]
    datatype: null;
};

export type IGraph = {
    type: 'Graph';
    value: IPSO[];
    datatype: null;
};

export async function parseN3File(file: string) : Promise<N3.Store> {
    logger.debug(`parsing: ${file}`);
    const n3 = fs.readFileSync(file, { encoding: "utf8"});
    return parseN3(n3);
}

export async function parseN3(n3: string) : Promise<N3.Store> {
    return new Promise<N3.Store>( (resolve,reject) => {
        const parser = new N3.Parser({ format: 'text/n3' });
        const store  = new N3.Store(); 

        parser.parse(n3,
            (error, quad, _prefixes) => {
                if (error) {
                    reject(error.message);
                }

                if (quad) {
                    store.add(quad)
                }
                else {
                    resolve(store);
                }
            }
        );
    });
}

export function serializeN3Store(store: N3.Store) : void {
    store.forEach( (quad) => {
        const subject   = quad.subject.value;
        const predicate = quad.predicate.value;
        const object    = quad.object.value;
        const graph     = quad.graph.value;

        console.log(`${subject} || ${predicate} || ${object} IN ${graph}`);
    }, null, null, null, null);
}

function pref(type: string, value: string) : string {
    return type + value;
}

export function writeDynamic(graph: IGraph, except: string[] = []) : string {
    const dynamicTerms = new Set<string>();

    graph.value.forEach( (pso) => {
        scanDynamicTerm(pso,dynamicTerms);
    });

    return  Array.from(dynamicTerms)
                 .filter( (dyn) => {
                    let result = true ;
                    except.forEach( (ex) => {
                        if (dyn.match(ex)) {
                            result = false;
                        }
                    });
                    return result;
                 })
                 .map( (dyn) => {
                        return `:- dynamic('<${dyn}>'/2).`;
                 }).join("\n");
}

function scanDynamicTerm(pso: IPSO, container: Set<string>) : void {
    if (pso.predicate.type === 'NamedNode') {
        container.add(pso.predicate.value);
    }

    if (pso.subject.type === 'Graph') {
        pso.subject.value.forEach( (pso_i) => {
            scanDynamicTerm(pso_i, container);
        });    
    }
    if (pso.predicate.type === 'Graph') {
        pso.predicate.value.forEach( (pso_i) => {
            scanDynamicTerm(pso_i, container);
        });    
    }
    if (pso.object.type === 'Graph') {
        pso.object.value.forEach( (pso_i) => {
            scanDynamicTerm(pso_i, container);
        });    
    }
}

export function writeGraph(graph: IGraph) : string {
    const result : string[] = [];

    graph.value.forEach( (pso) => {
        const value = writePSO(pso);
        result.push(value);
    });

    return result.join("\n");
};

function writePSO(pso: IPSO) : string {
    const subject   = writeTerm(pso.subject);
    const predicate = writeTerm(pso.predicate);
    const object    = writeTerm(pso.object);

    return `${predicate}(${subject},${object}).`;
}

function writeTerm(term: ITerm) : string {
    if (term.type === 'NamedNode') {
        return `'<${term.value}>'`;
    }
    else if (term.type === 'Literal') {
        if (term.datatype === pref(XSD,'string')) {
            return `'${term.value}'`;
        }
        else if (term.datatype === pref(XSD,'integer')) {
            return term.value;
        }
        else if (term.datatype === pref(XSD,'boolean')) {
            return term.value;
        }
        else {
            return `literal('${term.value}','${term.datatype}')`;
        }
    }
    else if (term.type === 'BlankNode') {
        return `'${term.value}'`;
    }
    else if (term.type === 'Variable') {
        return `${term.value}`;
    }
    else if (term.type === 'List') {
        const result : string[] = [];
        term.value.forEach( (li) => {
            result.push(writeTerm(li));
        });
        return '[' + result.join(",") + ']';
    }
    else if (term.type === 'Graph') {
        const result : string[] = [];
        term.value.forEach( (gi) => {
            const value = writePSO(gi);
            result.push(value.replace(/\.$/,''));
        });

        return '(' + result.join(",") + ')';
    }

    return 'x';
}

export function makeGraph(store: N3.Store, graph: N3.Term = N3.DataFactory.defaultGraph()) : IGraph {
    const result : IGraph = {
        type: 'Graph',
        value: [] as IPSO[],
        datatype: null
    };

    // First process the named nodes and literals...
    store.forEach((quad) => {
        const termType = '' + quad.subject.termType;

        if (termType === 'Variable') {
            console.error(quad);
            throw new Error(`Variables are not supported in N3S!`);
        }

        if ((termType === 'NamedNode' || termType === 'Literal')
                && !isGraphLike(quad,graph)) {
            let subject   = parseTerm(quad.subject, store);
            let predicate = parseTerm(quad.predicate, store);
            let object    = parseTerm(quad.object, store);
            result.value.push({
                type: 'PSO',
                subject: subject,
                predicate: predicate,
                object: object
            } as IPSO);
        }
    }, null, null, null, graph);

    // Next process the explicit bnodes...
    store.forEach((quad) => {
        const termType = '' + quad.subject.termType;
        if (termType === 'BlankNode' 
                && !isListLike(quad) 
                && !isGraphLike(quad,graph)) {
            let subject   = parseTerm(quad.subject, store);
            let predicate = parseTerm(quad.predicate, store);
            let object    = parseTerm(quad.object, store);
            result.value.push({
                type: 'PSO', 
                subject: subject,
                predicate: predicate,
                object: object
            } as IPSO);
        }
    }, null, null, null, graph);

    // Next process all the rest ...
    store.forEach((quad) => {
        const termType = '' + quad.subject.termType;
        if (termType === 'BlankNode' 
                && isListLike(quad) 
                && !isGraphLike(quad,graph)) {
            let subject   = parseTerm(quad.subject, store);
            let predicate = parseTerm(quad.predicate, store);
            let object    = parseTerm(quad.object, store);
            result.value.push({
                type: 'PSO', 
                subject: subject,
                predicate: predicate,
                object: object
            } as IPSO);
        }
    }, null, null, null, graph);

    return result;
}

function parseTerm(term: N3.Term, store: N3.Store) : ITerm {
    if (term.termType === 'NamedNode') {
        if (term.value === pref(RDF,'nil')) {
            return { type: 'List' , value: [] as ITerm[] } as IList;
        }
        else {
            return { type: 'NamedNode' , value: term.value} as INamedNode;
        }
    }
    else if (term.termType === 'Literal') {
        return {
            type: 'Literal',
            value: term.value ,
            datatype: term.datatypeString
        } as ILiteral;
    }
    else if (term.termType === 'BlankNode') {
        if (isList(term,store)) {
            return makeList(term,store);
        }
        else if (isGraph(term,store)) {
            return makeGraph(store,term);
        }
        else {
            const genid = makeGenId(term);
            return {
                type: 'BlankNode',
                value: genid
            } as IBlankNode;   
        }
    }
    else if (term.termType === 'Variable') {
        return {
            type: 'Variable',
            value: term.value
        } as IVariable;
    }
    else {
        return {
            type: 'BlankNode',
            value: 'unknown'
        } as IBlankNode;
    }
}

function makeGenId(term: N3.Term) : string {
    const value = term.value.replace(/^.*(_|\.)/,'') ;
    return `_:${value}`;   
}

function isGraphLike(quad: N3.Quad, graph: N3.Term) : boolean {
    if (quad.graph.id === graph.id) {
        return false;
    }
    else {
        return true;
    }
}

function isListLike(quad: N3.Quad) : boolean {
    if (quad.predicate.value === pref(RDF,'first') ||
        quad.predicate.value === pref(RDF,'rest')) {
        return true;
    }
    else {
        return false;
    }
} 

function isList(term: N3.Term, store: N3.Store) : boolean {
    let searchTerm = term;
    let brake = false;
    do {
        const first = store.getQuads(searchTerm,pref(RDF,'first'),null,null);
        const rest = store.getQuads(searchTerm,pref(RDF,'rest'),null,null);

        if (first.length == 1 && rest.length == 1) {
            // we are ok
        }
        else {
            return false;
        }

        if (rest[0].object.value === pref(RDF,'nil')) {
            brake = true;
        }
        else {
            searchTerm = rest[0].object;
        }
    } while (!brake);

    return true;
}

function isGraph(term: N3.Term, store: N3.Store) : boolean {
    const graph = store.getQuads(null, null, null, term);

    if (graph.length == 0) {
        return false;
    }
    else {
        return true;
    }
}

function makeList(term: N3.Term, store: N3.Store) : IList {
    let termList : ITerm[] = [];
    let searchTerm = term;
    let brake = false;

    do {
        const first = store.getQuads(searchTerm,pref(RDF,'first'),null,null);
        const rest  = store.getQuads(searchTerm,pref(RDF,'rest'),null,null);

        if (first.length == 0) {
            if (rest.length == 0 || rest.length != 1) {
                brake = true;
            }
            else {
                brake = true;
            }
        }
        else if (first.length != 1 || rest.length != 1) {
            brake = true;
        } 
        else if (first[0].object.value === pref(RDF,'nil')) {
            const termValue = { type: 'List', value: [] as ITerm[]} as IList;

            termList.push(termValue);

            if (rest[0].object.value === pref(RDF,'nil')) {
                brake = true;
            }
            else {
                searchTerm = rest[0].object;
            }
        }
        else {
            const termValue = parseTerm(first[0].object, store);

            termList.push(termValue);

            if (rest[0].object.value === pref(RDF,'nil')) {
                brake = true;
            }
            else {
                searchTerm = rest[0].object;
            }
        }

        first.forEach( (quad) => { store.removeQuad(quad) });
        rest.forEach( (quad) => { store.removeQuad(quad) });
    } while (!brake);
    
    return { type: 'List', value: termList } as IList;
}

src/N3SLexer.ts
export const XSD  = 'http://www.w3.org/2001/XMLSchema#';

export const xsd =  {
    decimal: `${XSD}decimal`,
    boolean: `${XSD}boolean`,
    double:  `${XSD}double`,
    integer: `${XSD}integer`,
    string:  `${XSD}string`,
  };

export type IN3SToken = {
    type: string;
    value: string;
    prefix: string;
    line: number;
    start: number;
    end: number;
};

export class N3SLexer {
    private input : string;
    private endOfFile : RegExp;
    private simpleApostropheString: RegExp;
    private unescapedIri : RegExp;
    private blank : RegExp;
    private boolean : RegExp;
    private literal : RegExp;
    private number : RegExp;
    private newline : RegExp;
    private comment : RegExp;
    private whitespace : RegExp;
    private directive: RegExp;
    private line : number;
    private comments : boolean;

    constructor(options?: any) {
        this.input = "";
        this.line = 0;
        this.simpleApostropheString = /^'([^']+)'/;
        this.unescapedIri = /^'<([^\x00-\x20<>\\"\{\}\|\^\`]*)>'/;
        this.blank = /^'_:((?:[0-9A-Z_a-z\xc0-\xd6\xd8-\xf6\xf8-\u02ff\u0370-\u037d\u037f-\u1fff\u200c\u200d\u2070-\u218f\u2c00-\u2fef\u3001-\ud7ff\uf900-\ufdcf\ufdf0-\ufffd]|[\ud800-\udb7f][\udc00-\udfff])(?:\.?[\-0-9A-Z_a-z\xb7\xc0-\xd6\xd8-\xf6\xf8-\u037d\u037f-\u1fff\u200c\u200d\u203f\u2040\u2070-\u218f\u2c00-\u2fef\u3001-\ud7ff\uf900-\ufdcf\ufdf0-\ufffd]|[\ud800-\udb7f][\udc00-\udfff])*)(?:[ \t]+|(?=\.?[,;:\s#()\[\]\{\}"'<>]))'/;
        this.literal = /^literal\(([']?[^',]+[']?),'([^']+)'\)/;
        this.number = /^[\-+]?(?:(\d+\.\d*|\.?\d+)[eE][\-+]?|\d*(\.)?)\d+/;
        this.endOfFile = /^(?:%[^\n\r]*)?$/;
        this.boolean = /^(?:true|false)/;
        this.newline = /^[ \t]*(?:%[^\n\r]*)?(?:\r\n|\n|\r)[ \t]*/;
        this.comment = /^%([^\n\r]*)/;
        this.directive = /^:-[ \t]*([^\n\r]*)/;
        this.whitespace = /^[ \t]+/;
        this.comments = false;

        if (options?.comments) {
            this.comments = true;
        }
    }

    public tokenize(input:string, callback?: (e: Error|null, token: IN3SToken)=>void) {
        this.input = this.readStartingBom(input);
        const tokens : IN3SToken[] = [];

        if (callback) {
            this.tokenizeToEnd(callback);
            return null;
        }
        else {
            let error;
            this.tokenizeToEnd((e: Error|null, t:IN3SToken) => {
                return e ? (error = e) : tokens.push(t);
            });
            if (error) throw error;
            return tokens;
        }
    }

    private readStartingBom(input: string) : string {
        return input.startsWith('\ufeff') ? input.substr(1) : input;
    }

    private tokenizeToEnd(callback: (e:Error|null,t:IN3SToken)=>void) : void {
        let input = this.input;
        let currentLineLength = input.length;

        while (true) {
            // Count and skip whitespace lines
            let whiteSpaceMatch, comment;

            while (whiteSpaceMatch = this.newline.exec(input)) {
                // Try to find a comment
                if (this.comments && (comment = this.comment.exec(whiteSpaceMatch[0]))) {
                  emitToken('comment', comment[1], '', this.line, whiteSpaceMatch[0].length);
                }
                // Advance the input
                input = input.substr(whiteSpaceMatch[0].length, input.length);
                currentLineLength = input.length;
                this.line++;
            }

            // Skip whitespace on current line
            if (!whiteSpaceMatch && (whiteSpaceMatch = this.whitespace.exec(input))) {
                input = input.substr(whiteSpaceMatch[0].length, input.length);
            }

            const line = this.line, firstChar = input[0];

            let type = '', value = '', prefix = '',
                match = null, matchLength = 0;

            if (this.endOfFile.test(input)) {
                emitToken('eof', '', '', this.line, 0);
                return;
            }

            switch(firstChar) {
                case ':':
                    if (match = this.directive.exec(input)) {
                        type = 'directive', value = match[1];
                    }
                    break;
                case '\'':
                    if (match = this.unescapedIri.exec(input)) {
                        type = 'IRI', value = match[1];
                    }
                    else if (match = this.blank.exec(input)) {
                        type = 'blank', prefix = '_', value = match[1];
                    }
                    else if (match = this.simpleApostropheString.exec(input)) {
                        type = 'literal' , value = match[1] , prefix = xsd.string;
                    }
                    break;
                case 'l':
                    if (match = this.literal.exec(input)) {
                        type = 'literal', value = match[1], prefix = match[2];
                    }
                    break;
                case 'f':
                case 't':
                    if (match = this.boolean.exec(input)) {
                        type = 'literal', value = match[0], prefix = xsd.boolean;
                    }
                    break;
                case '0':
                case '1':
                case '2':
                case '3':
                case '4':
                case '5':
                case '6':
                case '7':
                case '8':
                case '9':
                case '+':
                case '-':
                    // Try to find a number. Since it can contain (but not end with) a dot,
                    // we always need a non-dot character before deciding it is a number.
                    // Therefore, try inserting a space if we're at the end of the input.
                    if (match = this.number.exec(input)) {
                        type = 'literal', value = match[0];
                        prefix = (typeof match[1] === 'string' ? xsd.double :
                        (typeof match[2] === 'string' ? xsd.decimal : xsd.integer));
                    }
                    break;
                case '(':
                case ')':
                case '.':
                case ',':
                case '[':
                case ']':
                    matchLength = 1;
                    type = firstChar;
                    break;
                default:
                    matchLength = 1;
                    break;
            }

            if (!type) {
                return reportSyntaxError(this);
            }

            const length = matchLength || (match != null ? match[0].length : 0);

            emitToken(type,value,prefix,line,length);

            // Advance to next part to tokenize
            input = input.substr(length, input.length);
        }

        function emitToken(type: string, value: string, prefix: string, line: number, length: number) {
            const start = input ? currentLineLength - input.length : currentLineLength;
            const end = start + length;
            const token = { type, value, prefix, line, start, end } as IN3SToken;
            callback(null, token);
        }

        function reportSyntaxError(self: N3SLexer) { 
            let match = /^\S*/.exec(input);
            callback(self.syntaxError(match ? match[0] : ''),{} as IN3SToken); 
        }
    }

    // ### `syntaxError` creates a syntax error for the given issue
    private syntaxError(issue: string) {
        this.input = '';
        const err = new Error(`Unexpected "${issue}" on line ${this.line}.`);
        return err;
    }
}

src/N3SParser.ts
import { IN3SToken, N3SLexer } from "./N3SLexer";
import { IBlankNode, ILiteral, INamedNode, ITerm } from "./N3Parser";

type ICallBack = (token: IN3SToken) => ICallBack ;

export class N3SParser {
    private subject : ITerm;
    private lexer : N3SLexer;
    private readCallback : ICallBack;

    constructor() {
        this.lexer = new N3SLexer();
        this.subject = {} as ITerm;
        this.readCallback = this.readInTopContext;
    }

    private readInTopContext(token: IN3SToken) : ICallBack {
        switch(token.type) {
            case 'directive':
                // Ignore directives...
                return this.readSubject;
            default:
                return this.readSubject(token);
        }
    }

    private readSubject(token: IN3SToken) : ICallBack {
        switch (token.type) {
            case 'IRI':
                const entity = this.readEntity(token);
                if (entity !== undefined) {
                    this.subject = entity;
                }
                return this.readSubject;
            default:
                return this.readSubject;
        };
    }

    private readEntity(token: IN3SToken) : ITerm | undefined {
        let value = undefined;
        switch (token.type) {
            case 'IRI':
                value = {
                    type: 'NamedNode',
                    value: token.value,
                    datatype: null
                } as INamedNode;
                break;
            case 'literal':
                value = {
                    type: 'Literal',
                    value: token.value,
                    datatype: token.prefix
                } as ILiteral;
                break;
            case 'blank':
                value = {
                    type: 'BlankNode',
                    value: token.value,
                    datatype: null
                } as IBlankNode;
                break;
            default:
                this.error(`Expected entity but got ${token.type}`, token);
                value = undefined;
        };

        return value;
    }

    public parse(input: string) {
        this.lexer.tokenize(input)?.every( (token: IN3SToken) => {
            return this.readCallback = this.readCallback(token);
        });
        console.log(this.subject);
    }

    // ### `error` emits an error message through the callback
    private error(message: string, token: IN3SToken) {
        throw new Error(`${message} on line ${token.line}.`);
    }
}

src/lexer.ts
import { N3SParser } from './N3SParser';

main();

async function main() : Promise<void> {
  //  const lexer = new N3.Lexer({ lineMode: false , n3: true});

    //const sample = "'<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('<urn:example.org:Alice>','<urn:example.org:Person>').";
    //const sample = "'_:x'(true,'<urn:example.org:Person>').";
    //const sample = "'_:x'(false,'<urn:example.org:Person>').";
    //const sample = "'_:x'(12.3,'<urn:example.org:Person>').";
    //const sample = "'_:x'(false,literal('3.14','http://www.w3.org/2001/XMLSchema#float')).";
    //const sample  = "% test \n'<http://www.w3.org/2000/10/swap/log#onNegativeSurface>'(['_:X'],('<http://www.w3.org/2000/10/swap/log#onNegativeSurface>'([],('<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('_:X','<urn:example.org:Human>'))),'<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('_:X','<urn:example.org:Person>'))).";
    //const sample = "'_:x'(false,'hello world').\ntrue(true,true).";
    //const sample = "-12.23";
    //const sample = ".";
    //const sample = "% a comment\n";
    const sample = ":- dynamics foo\n'<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'('<urn:example.org:Alice>','<urn:example.org:Person>').";

   // lexer.tokenize(sample).every( token => {
//        console.log(token);
     //   return true;
//    });

    // const myLexer = new N3SLexer();

    // myLexer.tokenize(sample)?.every( token => {
    //     console.log(token);
    //     return true;
    // });

    const parser = new N3SParser();

    parser.parse(sample);
}